File: backend/src/domain/keys.ts
/**
 * Single-table layout
 * ===================
 * Every tenant (a library branch / tool room) is one partition, so all reads
 * for a branch stay inside one partition key and cost one Query.
 *
 *  Entity        PK                SK                     GSI1 (open loans)          GSI2 (lookup)                 GSI3 (history)
 *  -----------------------------------------------------------------------------------------------------------------------------
 *  Org           ORG#<org>         ORG                    -                          ORGS / ORG#<org>              -
 *  Member        ORG#<org>         MEMBER#<id>            -                          ORG#<org>#PHONE#<e164>        -
 *  Asset         ORG#<org>         ASSET#<id>             -                          ORG#<org>#CODE#<code>         -
 *  Checkout      ORG#<org>         CHECKOUT#<id>          ORG#<org>#OPEN / <dueAt>   -                             ORG#<org>#MEMBER#<id>
 *  Notification  ORG#<org>         NOTIF#<id>             -                          -                             ORG#<org>#MEMBER#<id>
 *
 * GSI1 is sparse: the attributes are removed on check-in, so it only ever holds
 * currently borrowed items sorted by due date. That makes both the overdue
 * dashboard and the nightly reminder sweep a single ranged Query.
 */

export const PK = (orgId: string) => `ORG#${orgId}`;

export const SK = {
  org: () => 'ORG',
  member: (memberId: string) => `MEMBER#${memberId}`,
  asset: (assetId: string) => `ASSET#${assetId}`,
  checkout: (checkoutId: string) => `CHECKOUT#${checkoutId}`,
  notification: (notificationId: string) => `NOTIF#${notificationId}`,
};

export const PREFIX = {
  member: 'MEMBER#',
  asset: 'ASSET#',
  checkout: 'CHECKOUT#',
  notification: 'NOTIF#',
};

export const GSI1 = 'gsi1';
export const GSI2 = 'gsi2';
export const GSI3 = 'gsi3';

export const openLoansPk = (orgId: string) => `ORG#${orgId}#OPEN`;
/** dueAt first so the range key sorts by urgency; id keeps it unique. */
export const openLoansSk = (dueAt: string, checkoutId: string) => `${dueAt}#${checkoutId}`;

export const ORG_REGISTRY_PK = 'ORGS';
export const phoneLookupPk = (orgId: string, phone: string) => `ORG#${orgId}#PHONE#${phone}`;
export const assetCodeLookupPk = (orgId: string, code: string) =>
  `ORG#${orgId}#CODE#${code.toUpperCase()}`;

export const memberHistoryPk = (orgId: string, memberId: string) =>
  `ORG#${orgId}#MEMBER#${memberId}`;

export function idFromSk(sk: string, prefix: string): string {
  return sk.startsWith(prefix) ? sk.slice(prefix.length) : sk;
}

File: backend/src/domain/keys.py
def pk(org_id: str) -> str:
    return f"ORG#{org_id}"


def sk_org() -> str:
    return "ORG"


def sk_member(member_id: str) -> str:
    return f"MEMBER#{member_id}"


def sk_asset(asset_id: str) -> str:
    return f"ASSET#{asset_id}"


def sk_checkout(checkout_id: str) -> str:
    return f"CHECKOUT#{checkout_id}"


def sk_unit(asset_id: str, unit_id: str) -> str:
    return f"UNIT#{asset_id}#{unit_id}"


PREFIX_MEMBER = "MEMBER#"
PREFIX_ASSET = "ASSET#"
PREFIX_CHECKOUT = "CHECKOUT#"
PREFIX_UNIT = "UNIT#"

GSI1 = "gsi1"
GSI2 = "gsi2"
GSI3 = "gsi3"

ORG_REGISTRY_PK = "ORGS"


def open_loans_pk(org_id: str) -> str:
    return f"ORG#{org_id}#OPEN"


def open_loans_sk(due_at: str, checkout_id: str) -> str:
    return f"{due_at}#{checkout_id}"


def phone_lookup_pk(org_id: str, phone: str) -> str:
    return f"ORG#{org_id}#PHONE#{phone}"


def asset_code_lookup_pk(org_id: str, code: str) -> str:
    return f"ORG#{org_id}#CODE#{code.upper()}"


def member_history_pk(org_id: str, member_id: str) -> str:
    return f"ORG#{org_id}#MEMBER#{member_id}"


def unit_serial_lookup_pk(org_id: str, serial: str) -> str:
    return f"ORG#{org_id}#UNIT#{serial.upper()}"
